phylanx/util/matrix_iterators.hpp

```diff
@@ -1,17 +1,46 @@
-//  Copyright (c) 2018 Parsa Amini
-//  Copyright (c) 2018 Hartmut Kaiser
+// Copyright (c) 2018 Parsa Amini
+// Copyright (c) 2018 Hartmut Kaiser
 //
-//  Distributed under the Boost Software License, Version 1.0. (See accompanying
-//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
 #if !defined(PHYLANX_MATRIX_ITERATORS)
 #define PHYLANX_MATRIX_ITERATORS
 
+#include <algorithm>
 #include <cstddef>
+#include <utility>
+
 #include <blaze/Math.h>
 
+namespace blaze
+{
+    // BADBAD: This overload of swap is necessary to work around the problems
+    //         caused by matrix_row_iterator not being a real random access
+    //         iterator. Dereferencing matrix_row_iterator does not yield a
+    //         true reference but only a temporary blaze::Row holding true
+    //         references.
+    //
+    // A real fix for this problem is proposed in PR0022R0
+    // (http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/p0022r0.html)
+    //
+    template <typename T>
+    HPX_FORCEINLINE
```

Comment:
I think this should not be added to the iterators, it's merely a facility that should be visible to the shuffle primitive (i.e. should be placed in shuffle_primitive.cpp).

Reply:
I think I got it right. Please check the updated code.